### PR TITLE
Improve #[cfg] diagnostics

### PR DESCRIPTION
**Pull request by jonas-schievink**
Unfortunately I ran into https://github.com/rust-analyzer/rust-analyzer/issues/4058 while testing this on https://github.com/nrf-rs/nrf-hal/, so I didn't see much of it in action yet, but it does seem to work.

### CODE REVIEW

File: crates/cfg/src/cfg_expr.rs [hunk 1]
```diff
@@ -2,30 +2,77 @@
...
+    /// For example, `#[cfg(target_feature = "sse")]` and `#[cfg(target_feature = "sse2")]`.
+    KeyValue { key: SmolStr, value: SmolStr },
+}
+
+impl CfgAtom {
+    /// Returns `true` when the atom comes from the target specification.
+    ///
+    /// If this returns `true`, then changing this atom requires changing the compilation target. If
+    /// it returns `false`, the atom might come from a build script or the build system.
+    pub fn is_target_defined(&self) -> bool {
```
Comment (jonas-schievink) [inline crates/cfg/src/cfg_expr.rs:26]: Note: Currently unused

File: crates/cfg/src/dnf.rs [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+
+        res.enabled.sort_unstable();
+        res.enabled.dedup();
+        res.disabled.sort_unstable();
+        res.disabled.dedup();
+        Some(res)
+    }
+
+    /// Returns `CfgDiff` objects that would enable this directive if applied to `opts`.
+    pub fn compute_enable_hints<'a>(
```
Comment (jonas-schievink) [inline crates/cfg/src/dnf.rs:68]: Note: Currently unused

File: crates/cfg/src/cfg_expr.rs [hunk 3]
```diff
@@ -2,30 +2,77 @@
...
+            ),
+        }
+    }
+}
+
+impl fmt::Display for CfgAtom {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CfgAtom::Flag(name) => write!(f, "{}", name),
+            CfgAtom::KeyValue { key, value } => write!(f, "{} = \"{}\"", key, value),
```
Comment (matklad) [inline crates/cfg/src/cfg_expr.rs:47]:
```suggestion
            CfgAtom::KeyValue { key, value } => write!(f, "{} = {:?}", key, value),
```

Don't want to open us to attribute injection attack :D

File: crates/hir_def/src/diagnostics.rs [hunk 4]
```diff
@@ -94,15 +96,23 @@ impl Diagnostic for UnresolvedImport {
...
         DiagnosticCode("inactive-code")
     }
     fn message(&self) -> String {
-        // FIXME: say *why* it is configured out
-        "code is inactive due to #[cfg] directives".to_string()
+        let inactive = DnfExpr::new(self.cfg.clone()).why_inactive(&self.opts);
+        let mut buf = "code is inactive due to #[cfg] directives".to_string();
+
+        if let Some(inactive) = inactive {
+            write!(buf, ": {}", inactive).unwrap();
```
Comment (matklad) [inline crates/hir_def/src/diagnostics.rs:112]:
```suggestion
            format_to!(buf, ": {}", inactive);
```

File: crates/cfg/src/dnf.rs [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+            write!(f, ")")?;
+        }
+
+        Ok(())
+    }
+}
+
+struct Conjunction {
+    literals: Vec<Literal>,
+}
```
Comment (matklad) [inline crates/cfg/src/dnf.rs:138]: As per https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#order-of-items, let's put Conjunction and Literal at the top

File: crates/cfg/src/dnf.rs [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+    }
+
+    /// Computes a list of present or absent atoms in `opts` that cause this expression to evaluate
+    /// to `false`.
+    ///
+    /// Note that flipping a subset of these atoms might be sufficient to make the whole expression
+    /// evaluate to `true`. For that, see `compute_enable_hints`.
+    ///
+    /// Returns `None` when `self` is already true, or contains errors.
+    pub fn why_inactive(&self, opts: &CfgOptions) -> Option<InactiveReason> {
```
Comment (matklad) [inline crates/cfg/src/dnf.rs:34]: I wonder if this should return `Result<(), InactiveReaason>`?

File: crates/cfg/src/dnf.rs (hunk 2 above)
Comment (matklad) [reply crates/cfg/src/dnf.rs:68]: I wonder how soon we will see `mod cdcl;`...

**Review by matklad — APPROVED** [review]

**Pull request merged**